docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+
+In Calcite, AST nodes are represented as subclasses of `SqlNode`. Column
+and type references are represented by `SqlIdentifier`. Literals are
+represented by `SqlLiteral` and its subclasses. Parameters are
+represented by `SqlDynamicParam`. Most other, if not all, expressions
+are subclasses of `SqlCall`.
+
+### Validation Stage
+
+The goal of the validation stage is to make sure that the AST
+expressions produced by the previous parsing stage are semantically
+valid. For instance, `1 + 1` is semantically valid while `1 + true` is
+not.
+
+In Calcite, the validation logic is encapsulated and orchestrated by
+implementations of `SqlValidator`. Specifically, `SqlValidatorImpl`
+extended by `HazelcastSqlValidator` is used to customize the validation
+logic to align it with our requirements.
+
+As a first step of the validation process, `SqlValidatorImpl` may apply
+some transformations to the passed AST which should not affect the
+semantics of the AST. Such transformations are called "unconditional
+rewrites" (see `SqlValidatorImpl.performUnconditionalRewrites`). Every
+`SqlCall` node has an associated `SqlOperator` returned by its
+`getOperator` method. `SqlOperator` implementations may customize
+various aspects of a certain operator including the rewriting behavior
+itself (see `SqlOperator.rewriteCall`).
+
+Normally, Calcite's parser inspects the hardcoded `SqlStdOperatorTable`
+implementation of `SqlOperatorTable` to resolve a certain operator, so
+it's hard to customize the pre-existing operators. To make the
+customization possible, operators from `SqlStdOperatorTable` are patched
+to the ones from `HazelcastSqlOperatorTable` during the AST rewriting
+process (see `HazelcastSqlValidator.performUnconditionalRewrites` and
+`HazelcastOperatorTableVisitor`).
+
+Usually, semantic validation requires assigning types to every
+expression and its operand expressions, if any. In Calcite, that type
+assignment behavior may be customized in several ways:
+
+- `SqlValidator.deriveType` may customize the behavior globally.
+
+- Each `SqlOperator` implementation may customize its operand type
+  inference strategy (see `SqlOperandTypeInference` and
+  `SqlOperator.getOperandTypeInference`) and its return type inference
+  strategy (see `SqlReturnTypeInference`,
+  `SqlOperator.getReturnTypeInference` and
+  `SqlOperator.inferReturnType`). The first one is used only if some of
+  the operand types are unknown. Also, operand type validation strategy
+  can be customized using an associated `SqlOperandTypeChecker` (see
+  `SqlOperator.getOperandTypeChecker`).
+
+- Additionally, some expressions require their operands to be coerced to
+  a certain type. For instance, `=` operator requires both of its sides
+  to be of the same type. In Calcite, such coercion services are
+  provided by `TypeCoercion` implementations. Specifically,
+  `TypeCoercionImpl` extended by `HazelcastTypeCoercion` is used to
```

Comment:
Can we have an extended description of how the coercion works (built-in and our own), how it relates to `01-type-system.md` conversions, what are the current problems, and possibly - how we are going to deal with it in the future?

Reply:
For example, we have `HazelcastTypeCoercion` that does some inference, and also `SqlOperandTypeInference` that passed to many operators, plus several custom implementations of this interface located in the `HazelcastInferTypes`. 
How do they relate to each other?

Reply:
clarified in d1d2929413835f7a462bc88686630dd57203b27c